src/main/java/org/privacyidea/AsyncRequestCallable.java

```diff
@@ -0,0 +1,90 @@
+package org.privacyidea;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.Response;
+import org.jetbrains.annotations.NotNull;
+
+import static org.privacyidea.PIConstants.ENDPOINT_AUTH;
+
+/**
+ * Instances of this class are submitted to the thread pool so that requests can be executed in parallel.
+ */
+public class AsyncRequestCallable implements Callable<String>, Callback {
+
+    private String path;
+    private final String method;
+    private Map<String, String> headers, params;
+    private final boolean authTokenRequired;
+    private final Endpoint endpoint;
+    private final PrivacyIDEA privacyIDEA;
+    final String[] callbackResult = {null};
+    private CountDownLatch latch;
+
+    public AsyncRequestCallable(PrivacyIDEA privacyIDEA, Endpoint endpoint, String path, Map<String, String> params,
+                                Map<String, String> headers, boolean authTokenRequired, String method) {
+        this.privacyIDEA = privacyIDEA;
+        this.endpoint = endpoint;
+        this.path = path;
+        this.params = params;
+        this.headers = headers;
+        this.authTokenRequired = authTokenRequired;
+        this.method = method;
+    }
+
+    @Override
+    public String call() throws Exception {
+        // If an auth token is required for the request get that first then do the actual request
+        if (this.authTokenRequired) {
```

Comment:
of course that would be possible but it would require even more logic for that. The auth token is also only required for /validate/triggerchallenge and /token/init (which should not be used anyway).